psydac/fem/vector.py

```diff
@@ -156,8 +155,19 @@ class ProductFemSpace( FemSpace ):
     Product of FEM space
     """
 
+    def __new__(cls, *args, **kwargs):
+
+        if len(args) == 1:
+            return args[0]
+        else:
+            return FemSpace.__new__(cls)
+
     def __init__( self, *args, **kwargs ):
         """."""
+
+        if len(args) == 1:
+            return
+
```

Comment:
Here I think we are in the class `ProductFemSpace`. But does that mean that in the case of a single argument, the constructor returns an object which will not be a ProductFemSpace ?

Reply:
I assumed so, but we do for some reason

Reply:
We do what? For which reason ?

Reply:
@campospinto You're right of course! This is class `ProductFemSpace`. In the linear algebra layer class `BlockVectorSpace` returns the input `VectorSpace` if only one is provided (see [here](https://github.com/pyccel/psydac/blob/79392da8a71f0c1aeaa11ccc48158c7aaaf3f37c/psydac/linalg/block.py#L14-L39)), so it is natural to do the same in the FEM layer.

Reply:
@saidctb Can you please read my original comment (the first at the top of this discussion) and reply to it?

Reply:
if you delete that line in __init__, the tests will not work anymore.
if you run the test `test_api_1d_compatible_spaces.py` you will get this
```python
test_api_1d_compatible_spaces.py:55: in run_system_1_1d_dir
    ah = discretize(equation, domain_h, [Xh, Xh])
../discretization.py:498: in discretize
    return DiscreteEquation(a, *args, **kwargs)
../discretization.py:114: in __init__
    self._lhs = discretize(expr.lhs, *newargs, **kwargs)
../discretization.py:489: in discretize
    return DiscreteBilinearForm(a, kernel_expr, *args, **kwargs)
../fem.py:157: in __init__
    quad_order                    = kwargs.pop('quad_order', get_quad_order(self.spaces[1]))
../fem.py:42: in get_quad_order
    return get_quad_order(Vh.spaces[0])
../fem.py:42: in get_quad_order
    return get_quad_order(Vh.spaces[0])
E   RecursionError: maximum recursion depth exceeded
!!! Recursion detected (same locals & position)
```
the problem with that is even if `len(args) == 1`  it will still pass by the `__init___`

Reply:
This is really surprising! According to the [Python documentation](https://docs.python.org/3/reference/datamodel.html#object.__new__):

> If __new__() does not return an instance of cls, then the new instance’s __init__() method will not be invoked.

If in this case the `__init__()` method is invoked, it means that `args[0]` is an object of type `ProductFemSpace`. I don't see why this should happen, and I wonder if we have a bug somewhere else.

Reply:
yes I have zero ideas, why it's doing that

Reply:
I've just checked, the problem is [here](https://github.com/pyccel/psydac/blob/d9b19a86f86f2c60ed38ef3cdedb8e49e38807b7/psydac/api/discretization.py#L444):
```python
Vh = ProductFemSpace(*g_spaces.values())
```
when `g_spaces.values()` could contain just one object of type `ProductFemSpace`.

Reply:
@saidctb I think that this is not a bug, just a strange way to organize the algorithm. Your code seems OK to me, thanks!